Fix type conflict with @types/react

hasparus opened the pull request:
_Connected issues: https://github.com/react-spring/react-three-fiber/issues/172, https://github.com/react-spring/react-three-fiber/issues/11, https://github.com/react-spring/react-three-fiber/issues/10_

## Problem

Described in https://github.com/react-spring/react-three-fiber/issues/172.

_react-three-fiber_ types conflict with _@types/react_ what forces users leveraging TypeScript type system to skip typechecking libraries.

## Main change

I have removed `audio` and `line` from IntrinsicElements interface.
_@types/react_ "was winning" and shadowing three-fiber `audio` and `line` definitions, so they weren't usable for users with types for React.
This is a small, but **breaking change** for people using _react-three-fiber_ and _react_ without _@types/react_, what is possible. I'm not sure how much of JavaScript projects would be affected, because [VSCode automatic type acquisition](https://code.visualstudio.com/docs/languages/javascript#_intellisense-is-not-working-for-external-libraries) can get _@types/react_ for JavaScript users.

I've copycatted the [lie from React Native](https://github.com/react-spring/react-three-fiber/issues/172#issuecomment-522887899) and added `components` export with values typed as `React.FC<...>`, which are strings at runtime.

- Using this makes react-three-fiber independent of global JSX.IntrinsicElements 
- Allows seamlessly adding actual components with render props etc.

```tsx
import { components as t } from 'react-three-fiber'

const _ = (
  <t.Mesh
    onUpdate={mesh => console.log(mesh.geometry)}
    onClick={() => console.log('click')}
    onPointerOver={() => console.log('hover')}
    onPointerOut={() => console.log('unhover')}>
    <t.BoxBufferGeometry attach="geometry" args={[1, 1, 1]} />
    <t.MeshNormalMaterial attach="material" />
  </t.Mesh>
)
```

It would be pretty cool to allow to
```tsx
import { Mesh } from 'react-three-fiber/components'`
```
but I didn't want to complicate the PR.

## More changes

- Added jest to unit test `components` a bit.

- Removed _@types/three_, because it's deprecated. THREE ships its own types.
  - Dev dependency on three will provide types during development.
  - Peer dependency will give them to the library user.
  - If a the library user uses THREE in version without types, they can install proper version of _@types/three_ themselves.

- Disabled '@typescript-eslint/no-empty-interface'. This rule doesn't know if you extend a mapped type which is the case for ThreeFiberComponents.

- `yarn typecheck` didn't work for me 😔
  - I added `--emitDeclarationOnly false` to make `noEmit` work.

## Notes

I [think](https://github.com/react-spring/react-three-fiber/compare/master...hasparus:react-types-conflict?expand=1#diff-501113a0a07cf9a1fafb7c5106513d50R22) `components.tsx` is a little bit more complicated than it has to be.
_We could map over THREE exports without losing type info (no Object.entries) and instead of type asserting entire dict `as any as ThreeFiberComponents`, assert string value to a function component with proper props._
I'm not sure if I'm capable enough to simplify it soon :c

Deployed demo with my changes:
  - https://disastrous-juice.surge.sh/
  - https://github.com/hasparus/disastrous-juice

hasparus commented on package.json:
```diff
@@ -60,8 +60,9 @@
     "resize-observer-polyfill": "^1.5.1",
-    "scheduler": "0.17.0",
-    "tiny-emitter": "^2.1.0"
+    "scheduler": "0.13.3",
```
I accidentally decreased the version on merge conflict.

drcmda commented:
thanks for this! it's lots to take in, so, the change here is that there's an optional way to get components and it isn't that much weight?

i would also prefer react-three-fiber/components btw i think that would make imports cleaner

hasparus commented:
> i would also prefer react-three-fiber/components btw i think that would make imports cleaner

That would also contribute to autoimport, so when I write <Mesh /> the text editor can propose me to import Mesh from _three_ or a Mesh from _react-three-fiber/components_.

> native elements are still working

Yup, I didn't touch the reconciler.

I removed _audio_ and _line_ from types. I've noticed that the editor support I get on them _prefers_ their declarations from _@types/react_. They still work at runtime though.
This change can fail CI build for somebody who's using react-three-fiber without @types/react.
It's kind of "breaking" I'd say, so it could probably be a part of next major release.

> it isn't that much weight

Ouch. It's good you've mentioned it.

The `components` object is just a string for every three export.

I imported _three_ (which is [huge](https://bundlephobia.com/result?p=three@0.110.0)) and ran Object.entries on it to generate the `components` export.
I didn't check yet, but this should break treeshaking on three.

**I'm not sure if this is a problem, though. Doesn't react-three-fiber imports entire three already?**
https://github.com/react-spring/react-three-fiber/blob/master/src/reconciler.tsx#L186

### Possible solutions:

This is a long shot, because I'm not sure if it's a problem.
I'll try to measure bundle size.

- Using a proxy to do the mapping from `Key` to `key`.
  **Cons:**
  - Higher runtime cost than reading from object?
  - This would make `import { Mesh } from 'react-three-fiber/components` not plausible.
- Generating this mapping at build-time.
  ```tsx
  // components.tsx
  export const Mesh = 'mesh' as any as ThreeFiberComponents['Mesh'];
  export const BoxBufferGeometry = 'boxBufferGeometry' as any as ThreeFiberComponents['BoxBufferGeometry'];
  export const MeshNormalMaterial = 'meshNormalMaterial' as any as ThreeFiberComponents['MeshNormalMaterial'];
  ```
  **Pros:**
    - Makes named imports of components trivial.

  **Cons:**
    - Contributes to bundle size and adds another build step.
    - Needs to be regenerated after changes of _three_ public API?

drcmda commented:
if it's under /components im ok with it, it wouldn't matter if it pulls the entirety of three because it doesn't treeshake anyway. the only way i am aware of is using aliases, this is what i do to compress my projects and that would work with these wrapper components as well.

one more question i have is about the extend function. how would that go?

currently:

```jsx
import { export } from 'react-three-fiber'
import { SpecialThing } from '...'
extend({ SpecialThing })

<specialThing />
```

hasparus commented:
> the only sore spot is
> Needs to be regenerated after changes of three public API?

Yeah, this is a big disadvantage. Especially because _three_ is at version 0.x since forever and [this page](https://github.com/mrdoob/three.js/wiki/Migration-Guide) can be kinda scary when you have a large project.

> that one is something i wanted to avoid. tying the lib to a particular three version, or being forced to track their release cycle is not ideal.

I totally agree with it. This is a selling point of react-three-fiber.



I think I can do something that enables named imports from `react-three-fiber/components` and doesn't introduce a hard dependency on particular three version.

> one more question i have is about the extend function. how would that go?

This could work too

```ts
import { extend } from 'react-three-fiber'
import { SpecialThing } from 'xyz'

/* extend could return "specialThing" with a fake wrapper component type */
const { SpecialThing: Special } = extend({ SpecialThing })

<Special />
```

hasparus commented:
> i've been thinking about another special object, like this:
>```tsx
> import { SpecialThing } from 'xyz'
>
> <new object={SpecialThing} args={[...]} {...props} />

I think you can't get intrinsic elements to be generic, so this wouldn't autocomplete/typecheck.

```tsx
import { New } from 'react-three-fiber'
import { SpecialThing } from 'xyz'

<New object={SpecialThing} args={[...]} {...props} />
```

Using a component for that (regardless if the logic is actually inside of it or in the reconciler) would be easier IMHO.

I've made a quick example in TypeScript playground https://links.typescript.social/5J6n

hasparus commented:
I'm gonna try to get named imports from `react-three-fiber/components` working tomorrow morning.

@codynova You might be interested in this PR.

hasparus commented:
> I'm gonna try to get named imports from react-three-fiber/components working tomorrow morning.

I can't use `module.exports` nor `export = `, because they [can't be compiled](https://babeljs.io/docs/en/babel-plugin-transform-typescript#caveats) to ESNext.

So if I'm not wrong, the only way to get named imports of object entries, would be to transform this object into a module at build time. \
(``` Object.entries(components).map(([k, v]) => `export const ${k} = ${v}`).join('\n') ```).

I see two problems with it.
- **a small one** — I *really* like this import syntax more. Especially, since it allows to import both entire namespace and particular imports at the same time (like `import React, { useState }`), but adding a build-time script for it is probably vanity.
- **a big one** — ties _react-three-fiber_ to one _three_ version.

drcmda commented:
and proxies? i mean, this would basically be an add-on anyway so people know what they're getting into. this wouldn't tie it too close to the main lib and we're not dependent to a threejs version. not sure about types, but you're the pros, it can probably be done in some way or another.

hasparus commented:
> and proxies? 

Well, since we can't export object as a module it makes no difference if it's computed at the startup of the app (like I do with `components` ATM) or on the go with a proxy. I'd guess that plain object would be better performance-wise.

Shipping without ES6 modules would be a regression and not worth the gain IMO.

> but you're the pros

By no means :D

drcmda commented:
so, if we have an optional react-three-fiber/components esm export, not tied to the main export. and it's primed to three@latest and we find someone that likes to update it regularly or leave it to the community to update it via pr's, would it then be good to go?

hasparus commented:
Yeah, I can add a script to package.json to dump what is right now components object into a components.tsx module so a PR to update it would be pretty easy to do.

Classes exported by _three_ don't change that often so a github action / a cron job would probably be a overkill.

drcmda commented:
yes, please, no cron jobs :-D it could just use the threefiber main three devdependency to generate this file yarn build.

hasparus commented:
I couldn't just print all `components` object entries as a module, because there are deprecated three exports (SplineCurve3, CubeGeometry) and things like CanvasRenderer with which I don't really know what to do. I feel that filtering them out in value world would be harder than excluding them in types, so I dumped the type instead.

I have also added `Primitive` fake component with a type that _probably_ matches applyProps behavior.

drcmda commented:
what's your take on this? should we merge? is it safe?

hasparus commented:
I've used it a little bit already, though I can't share the project. I'm gonna do the [pool game](https://dev.to/manan30/creating-a-rudimentary-pool-table-game-using-react-three-js-and-react-three-fiber-part-1-363l) tutorial using this and deploy the results this week. Maybe I'll find some problems along the way.

This is technically a breaking change for people using `audio` and `line` TypeScript without `@types/react` (I'm not sure how many of them exist, but I'd assume that at least 1?), so maybe it should go on `4.x`/`next` branch?

drcmda commented:
seriously, who's using our types without @types/react 🙂but i would be ok with it. any other ideas that could go into a major?

hasparus commented:
Well, I didn't finish my pool game, but I've fixed merge conflicts and adapted `/components` to the changes in package build. 

- https://priceless-ardinghelli-a980b6.netlify.com/
- https://github.com/hasparus/r3f-pool-game

hasparus commented on package.json:
```diff
@@ -1,12 +1,12 @@
   "name": "react-three-fiber",
   "private": true,
-  "version": "3.0.13",
+  "version": "4.0.0-rc.1338",
```
Oops. I didn't mean to commit this line..

drcmda commented:
im gonna add a v4 branch today, is it possible to target this one in this pr instead of master?

hasparus commented:
I've fixed the conflict and rebased to v4. I didn't have time to get deep into v4 changes, but I love what I've seen already. Especially the `Dom` component, because this is such a common use case.

drcmda commented:
@hasparus i am considering releasing v4, it's been in testing for a longer time now and the features are pretty good. should we take a longer time still for types or do you think it's all ready to go?

hasparus commented:
It should be okay, though I’d probably test it again after the merge with v4 just in case.
